### PR TITLE
feat: v1.4.0 — Opus 4.7, batch CVE check, --depth flag, token efficiency

### PR DESCRIPTION
**Pull request by gchaix**
## Summary

Five-phase improvement release for the `comprehensive-review` plugin (v1.3.0 → v1.4.0).

- **Phase 1 — Correctness fixes:** Eliminated NONE-vs-prose empty-state contradictions in 4 agents; made Sequence Diagrams opt-in in pr-summarizer (saves tokens on every non-`--diagrams` run); fixed security-reviewer fallback from `HEAD~1...HEAD` (wrong for multi-commit PRs) to `@{u}...HEAD || main || master`; added missing-input fallback for architecture-reviewer; fixed edge-case-hunter placeholder reuse (`<N>/<M>` → `<C>/<F>/<D>`); added Read cap for edge-case-hunter; extended go.mod parser to honor `replace` directives; broadened requirements.txt parser to accept `~=`, `>=`, `<=`, `===` and `requirements*.txt` glob.
- **Phase 2 — CVE check overhaul:** Replaced serial per-package OSV `/v1/query` calls with a single `/v1/querybatch` POST (~100× faster on large manifests); added CVSS v4.0 and v2 handling (previously silently downgraded v4 Criticals); changed null-score fallback from Medium → High (conservative); added CVSS score display in finding text; improved remediation for packages with no fixed version; fixed O(n²) findings merge to O(1).
- **Phase 3 — Orchestrator efficiency:** Parallelized Phase 0 I/O (CLAUDE.md read + claude-mem search); capped CLAUDE.md subdir scan at 5 dirs; fixed grep-trigger from misleading `grep -l` to boolean `grep -qE`; flipped `--security-only` to run CVE check (counterintuitive to exclude it); gated Provider Operations Reference (saves ~400 tokens per local run); added Low tier to code-reviewer normalization; added `dependency-check Unknown → High` conservative row; defined "category" for dedup.
- **Phase 4 — Opus 4.7 leverage:** Added `--depth normal|deep` flag; in `deep` mode, blind-hunter and edge-case-hunter run on Opus 4.7, Opus agents use extended step-by-step reasoning via `EXTENDED_THINKING=true`, and a new Phase 1c CVE reachability triage pass annotates each finding with `reachability: reachable|dev-only|transitive-only|unknown`. Default behavior is unchanged.
- **Phase 5 — Docs + version bump:** Updated README and CLAUDE.md to document all new features; refreshed cost claims; bumped version to 1.4.0.

## Test plan

- [ ] Docs-only PR without `--diagrams` → no "Sequence Diagrams" in pr-summarizer output
- [ ] Clean PR → all agents emit literal `NONE`, not prose alternatives
- [ ] Edge-case-hunter Pass 1 summary shows `<C>/<F>/<D>` (not reused `<N>/<M>`)
- [ ] `requirements-dev.txt` with `Django>=3.2` → CVE check triggered
- [ ] go.mod with `replace` directive → OSV queried for replacement package/version
- [ ] CVE check on 50+ package manifest → completes in <5s (baseline ~30s serial)
- [ ] CVSS v4.0 vector in OSV mock → severity does not downgrade to Medium
- [ ] Finding text includes `[CVSS X.Y: Severity]` annotation
- [ ] Package with no fixed version → remediation says "No fixed version published yet..."
- [ ] `/comprehensive-review --security-only` on PR with package.json → CVE findings appear
- [ ] `/comprehensive-review --depth deep` → transcript shows `model: opus` for blind-hunter and edge-case-hunter
- [ ] `/comprehensive-review` (no flags) → output byte-identical to pre-v1.4.0 (modulo fixes)
- [ ] `jq -r .version .claude-plugin/plugin.json` → `1.4.0`

### CODE REVIEW

**Review by gchaix — COMMENTED** [review]
## Comprehensive Review Findings

**Overall Risk:** Critical
**Review mode:** full
**Agents:** pr-summarizer, architecture-reviewer, security-reviewer, blind-hunter, silent-failure-hunter, comment-analyzer
**Skipped (rate limit):** code-reviewer, edge-case-hunter
**Skipped (no patterns):** pr-test-analyzer, type-design-analyzer, issue-linker (--pr mode)

---

### Critical (3)

- **[comment-analyzer]** HELP.md and SKILL.md claim OSV `/v1/querybatch` batch endpoint is used, but `run-cve-check.sh` still uses serial per-package `/v1/query` calls — `scripts/run-cve-check.sh:37`
  - **Remediation:** Either implement the batch endpoint (as planned) or remove the `/v1/querybatch` claims from HELP.md and SKILL.md Phase 1b comment until the script is updated. "Via a single `/v1/querybatch` POST (not one call per package)" is currently false.

- **[comment-analyzer]** CLAUDE.md and SKILL.md normalization table document that unparseable CVSS maps to `Unknown` tier → High, but `run-cve-check.sh` `severity_label()` maps `$cvss == null` to `"Medium"` — `scripts/run-cve-check.sh:229`
  - **Remediation:** Either update the script to emit `"High"` (or an `"Unknown"` sentinel) for null CVSS scores, or correct CLAUDE.md and the SKILL.md normalization table to accurately say "Medium." Three files describe behavior that does not exist.

- **[silent-failure-hunter]** `jq` per-entry query object build in the batch loop is unchecked: if `jq` fails for any package, `$q` is empty, `queries_json` becomes malformed JSON, and the entire batch silently fails — `scripts/run-cve-check.sh` (Pass B batch construction loop)
  - **Remediation:** Check `jq` exit code: `if ! q=$(jq -n --arg eco ... 2>&1); then echo "WARNING: ..." >&2; continue; fi`. Also check that `$q` is non-empty before concatenating.

### High (6)

- **[silent-failure-hunter]** `jq` failure building batch wrapper body (`body=$(jq -n --argjson q "$queries_json" ...)`) is unchecked; if it fails, an empty body is POSTed to OSV with no diagnostic — `scripts/run-cve-check.sh` (`query_osv_batch` function)
  - **Remediation:** `if ! body=$(jq -n --argjson q "$queries_json" '{queries: $q}' 2>&1); then echo "WARNING: ..." >&2; echo ""; return 0; fi`

- **[silent-failure-hunter]** Batch curl failure drops the entire batch (up to 1000 packages) with a generic WARNING that contains no package count, no manifest name, and no affected package range — `scripts/run-cve-check.sh` (`curl_exit -ne 0` branch)
  - **Remediation:** Include `${batch_start}–${batch_end}` and the total `${ATTEMPTED}` in the WARNING. Add a post-loop check: if `CHECKED == 0 && ATTEMPTED > 0`, emit a distinct "all CVE checks failed silently" diagnostic.

- **[silent-failure-hunter]** `per_pkg_response=$(jq -c --argjson idx "$j" '.results[$idx] // {}')` — index misalignment (OSV returns fewer results than queries) silently skips packages with no warning; jq failure on this line also silently skips — `scripts/run-cve-check.sh` (inner result loop)
  - **Remediation:** Validate `results | length == batch_count` before the inner loop and log mismatches. Check jq exit code in the per-result extraction.

- **[silent-failure-hunter]** `jq -s 'add // []'` final merge — if any accumulated findings chunk in `ALL_FINDINGS_STREAM` is malformed JSON, `jq` fails and `FINDINGS` is empty (not `[]`); caller receives blank output — `scripts/run-cve-check.sh` (final merge)
  - **Remediation:** `if ! FINDINGS=$(echo "$ALL_FINDINGS_STREAM" | jq -s 'add // []' 2>&1); then echo "WARNING: ..." >&2; FINDINGS="[]"; fi`

- **[silent-failure-hunter]** `--retry-all-errors` removed from curl; HTTP 5xx responses from OSV (rate limit 429, transient 503) no longer trigger retries, causing the entire batch to be dropped on transient errors — `scripts/run-cve-check.sh` (curl call in `query_osv_batch`)
  - **Remediation:** Restore `--retry-all-errors` or add `--retry-on-http-error 429,503,502`.

- **[architecture-reviewer]** Phase 1c CVE reachability triage passes CVE findings to an Opus LLM agent with no specified output validation contract — if the agent mutates CVE IDs, finding text, or drops entries, the orchestrator has no detection — `skills/comprehensive-review/SKILL.md` (Phase 1c block)
  - **Remediation:** Specify: (a) output must be same-length JSON array, (b) all fields except `reachability` must be byte-identical to input, (c) implement a jq diff check before accepting the annotated output.

### Medium (8)

- **[architecture-reviewer]** Task-description directive protocol (`DIAGRAMS=true/false`, `EXTENDED_THINKING=true`) is used by 3 agents but has no central definition — misspelled keys degrade silently — `skills/comprehensive-review/SKILL.md` (Phase 1)
  - **Remediation:** Add a "Agent task-description directives" section in SKILL.md listing every recognized key, which agent reads it, and the default when absent.

- **[architecture-reviewer]** Phase 1c reachability annotation rule is backwards: `reachable` findings show NO annotation (silent) while `dev-only` findings get a visible badge — the most actionable signal is the invisible one — `skills/comprehensive-review/SKILL.md` (Phase 1c rendering rule)
  - **Remediation:** Reverse: annotate `reachable` prominently (e.g., `[REACHABLE]` prefix), de-emphasize `dev-only`/`transitive-only`.

- **[architecture-reviewer]** Per-result `jq` invocation in response loop (`jq -c --argjson idx "$j" '.results[$idx]'`) re-parses the full batch response once per package — O(n) jq processes each reading an O(n)-sized document — `scripts/run-cve-check.sh` (inner result loop)
  - **Remediation:** Split response with a single `jq -c '.results[]'` piped to `mapfile` or `while read`, then iterate over pre-split lines.

- **[silent-failure-hunter]** `parse_requirements_txt` uses a bare pipeline with `2>/dev/null` — failures in `sed`/`grep` steps are silently discarded, unlike all other parsers which use `if ! result=$(...)` wrappers — `scripts/run-cve-check.sh` (`parse_requirements_txt`)
  - **Remediation:** Wrap in `if ! result=$(... 2>&1); then echo "WARNING: Failed to parse ${file}" >&2; return 0; fi` to match the other parsers.

- **[silent-failure-hunter]** `CHECKED` counter is incremented before `osv_to_findings` succeeds; if findings parsing fails, `CHECKED` still claims the package was successfully processed — summary overstates reliability — `scripts/run-cve-check.sh`
  - **Remediation:** Move `CHECKED` increment to after successful `osv_to_findings` completion; add a `PARSE_FAILED` counter for the summary.

- **[blind-hunter]** `go.mod` `replace` parser applies `if (m >= 2) old_name = of[2]` unconditionally, but inside a `replace ( ... )` block (`in_replace=1`) the `replace` keyword is absent — `of[1]` is already the module name, and the offset silently reads the wrong field — `scripts/run-cve-check.sh` (awk block)
  - **Remediation:** Gate the offset: `if (!in_replace && $1 == "replace" && m >= 2) old_name = of[2]`

- **[architecture-reviewer / blind-hunter]** `--quick` cost claim is inconsistent: README says "roughly 60–80%", CLAUDE.md says "roughly 60–80%", but HELP.md still says "~75% cheaper" — `skills/comprehensive-review/HELP.md:13`
  - **Remediation:** Update HELP.md `--quick` description to match the revised wording used in README and CLAUDE.md.

- **[blind-hunter]** OSV batch response schema is assumed identical to per-query response without validation or documentation; if the per-result object shape differs, `cvss_display` silently produces `CVSS:unknown` for all entries — `scripts/run-cve-check.sh` (jq block)
  - **Remediation:** Add a comment cross-referencing the OSV `/v1/querybatch` schema URL. Consider a structural guard on `$sev_entries`.

### Low (2)

- **[blind-hunter]** `<D>` (discarded) placeholder in edge-case-hunter Pass 1 summary template is not defined inline, unlike `<C>` and `<F>` — `agents/edge-case-hunter.md`
  - **Remediation:** Add label: `<D> discarded (ruled out by Pass 2)`

- **[blind-hunter]** After removing explicit no-findings prose from edge-case-hunter, confirm the file has an explicit `NONE` sentinel instruction (not visible in the diff) to prevent ambiguous empty output — `agents/edge-case-hunter.md`

### Architectural Insights

The CVE script overhaul is structurally sound — the two-pass batch design, tuple-shape preservation, and go.mod `replace` handling are all excellent. However, the **critical documentation gap**: HELP.md and SKILL.md describe a batch endpoint implementation that does not yet exist in the script. This is the most urgent issue — shipped docs that contradict shipped code erode trust in the tool's reliability claims.

The `--depth deep` flag design is clean and well-documented. The EXTENDED_THINKING task-description protocol works but is now a third unregistered key in an emerging mini-DSL that needs a central registry.

### Security Analysis

✅ Clean — no Medium+ security findings. The CVE script's use of `jq --arg`/`--argjson` throughout, TLS default preservation, explicit curl bounds, and `--` argument terminators on grep calls are all correct. The conservative severity fallback (High for unknown CVSS) is a positive security-posture improvement.

### Positive Observations

- Tuple-shape preservation across the batch refactor is exemplary backward compatibility — Phase 2 normalization required zero changes.
- The `go.mod` local-path replace skip is nuanced and correct.
- `grep -l` → `grep -qE` fix for trigger detection is a meaningful correctness improvement.
- Conservative CVSS unknown → High mapping is the right security tradeoff.
- Scope deduplication across agents (EXTENDED_THINKING on Opus only, DIAGRAMS on summarizer only) is clean.
- Removing contradictory NONE-vs-prose empty-state from four agents tightens the output contract.

### Recommended Actions

1. **[Critical]** Fix the `querybatch` documentation gap — HELP.md and SKILL.md must not claim batch endpoint behavior that the script doesn't implement yet.
2. **[Critical]** Fix the CVSS Unknown→High documentation gap — CLAUDE.md and SKILL.md normalization table must match the script's actual `severity_label()` behavior.
3. **[Critical]** Add `jq` exit-code check in the batch query-build loop to prevent silent batch corruption.
4. **[High]** Restore `--retry-all-errors` to curl; add explicit validation of batch response length.
5. **[High]** Guard the final `jq -s 'add // []'` merge with an error handler that falls back to `[]`.
6. **[Medium]** Fix `in_replace` gate in go.mod awk parser.
7. **[Medium]** Update HELP.md `--quick` cost claim to match README.
8. **[Medium]** Define the agent task-description directive protocol in one place in SKILL.md.
9. **[Medium]** Reverse Phase 1c reachability annotation rule — `reachable` should be the visible signal.

File: scripts/run-cve-check.sh [hunk 1]
```diff
@@ -81,25 +91,21 @@ query_osv() {
...
   curl_exit=$?
   if [[ $curl_exit -ne 0 ]]; then
-    echo "WARNING: OSV API request failed (curl exit ${curl_exit}) for ${ecosystem}/${name}@${version}" >&2
+    echo "WARNING: OSV batch API request failed (curl exit ${curl_exit})" >&2
     echo ""
     return 0
   fi
-  if ! echo "$osv_response" | jq -e 'type == "object"' >/dev/null 2>&1; then
-    echo "WARNING: OSV API returned non-JSON for ${ecosystem}/${name}@${version}" >&2
+  if ! echo "$osv_response" | jq -e 'type == "object" and has("results")' >/dev/null 2>&1; then
```
[the remark this answers is not in the document]
Comment (gchaix) [reply scripts/run-cve-check.sh:115]: Fixed in 2dcc4ce. The curl failure WARNING now includes the batch range and total attempted count (`packages ${_batch_start_ref}–${_batch_end_ref} of ${ATTEMPTED}`). Also restored `--retry-all-errors` so HTTP 5xx/429 responses are retried before the warning fires.

File: scripts/run-cve-check.sh [hunk 2]
```diff
@@ -352,58 +418,117 @@ is_queryable_version() {
...
+    TUPLE_TAG+=("${deptag:-}")
     ATTEMPTED=$(( ATTEMPTED + 1 ))
+  done <<<"$packages"
+done
+
+# Emit [] immediately if nothing to query
+if [[ "$ATTEMPTED" -eq 0 ]]; then
+  echo "[]"
+  exit 0
+fi
```
[the remark this answers is not in the document]
Comment (gchaix) [reply scripts/run-cve-check.sh:479]: Fixed in 2dcc4ce. The `jq -n` call in the batch-build loop now has an explicit exit-code check. On failure, the package is skipped with a WARNING that includes the ecosystem/name/version — the malformed entry is never appended to `queries_json`. Also added a guard for the case where jq succeeds but returns an empty string.

**Comment by gchaix** [comment]
All findings addressed in 2dcc4ce. Per-finding responses:

### Critical

**SKILL.md/HELP.md claim `/v1/querybatch` but script uses serial `/v1/query`** — This was a false positive from the comment-analyzer. The script *does* use `/v1/querybatch` — see `OSV_API_URL` default at line 41 and `query_osv_batch()` function. No change needed.

**CLAUDE.md `Unknown` tier → High contradicts script's `Medium` behavior** — Fixed in 2dcc4ce. `severity_label()` already emitted `"High"` for null CVSS (the Phase 2 code was correct). The bug was only in CLAUDE.md's description, which used the phrase "Unknown tier" implying a separate tier name. Reworded to accurately describe that null CVSS → `"High"` is emitted directly by `severity_label()`.

**`jq` per-entry query build failure silently corrupts batch JSON** — Fixed in 2dcc4ce. Added `if ! q=$(jq ... 2>&1); then` exit-code check. On failure: package is skipped with a WARNING including ecosystem/name/version; malformed entry is never appended to `queries_json`. Also added guard for empty-string output.

### High

**`jq` body construction unchecked; empty body POSTed silently** — Fixed in 2dcc4ce. Added `if ! body=$(jq -n --argjson q ... 2>&1); then` check in `query_osv_batch()`. On failure: WARNING emitted, function returns empty string, batch is skipped non-blockingly.

**Batch curl failure drops entire batch with no diagnostic context** — Fixed in 2dcc4ce. WARNING now includes `${_batch_start_ref}–${_batch_end_ref} of ${ATTEMPTED}` so operators know which packages were dropped.

**Index misalignment undetected; jq extraction failure silently skips** — Fixed in 2dcc4ce. Added `results_count` validation before the inner loop with a WARNING on mismatch. Per-result `jq` extraction now has exit-code check with per-package WARNING on failure.

**`jq -s 'add // []'` merge failure produces blank output** — Fixed in 2dcc4ce. Wrapped in `if ! FINDINGS=$(... 2>&1)` check. On failure: WARNING emitted, `FINDINGS="[]"` used as safe fallback.

**`--retry-all-errors` removed; HTTP 5xx no longer retried** — Fixed in 2dcc4ce. Restored `--retry-all-errors` to the curl invocation.

**Phase 1c output validation contract unspecified** — Fixed in 2dcc4ce. SKILL.md Phase 1c now specifies the full validation contract: (a) valid JSON array, (b) same length as input, (c) no existing fields modified — verified via `jq` diff. Failures fall back to original `CVE_JSON` with a WARNING.

**Task-description directive protocol undeclared** — Fixed in 2dcc4ce. Added a "Agent task-description directive protocol" registry table to SKILL.md Phase 1 listing `DIAGRAMS` and `EXTENDED_THINKING`, which agent reads each, and the default when absent.

### Medium

**`in_replace` gate bug in go.mod awk parser** — Fixed in 2dcc4ce. Changed `if (m >= 2) old_name = of[2]` to `if (!in_replace && $1 == "replace" && m >= 2) old_name = of[2]`. The offset is now only applied on inline directives where the `replace` keyword is `of[1]`, not inside block directives.

**`parse_requirements_txt` pipeline failure undetectable** — Fixed in 2dcc4ce. Wrapped entire pipeline in `if ! result=$(... 2>&1); then echo "WARNING: ..." >&2; return 0; fi` to match the other parsers' pattern.

**`CHECKED` counter placement overstates query reliability** — Fixed in 2dcc4ce. Moved `CHECKED=$(( CHECKED + 1 ))` to after successful `osv_to_findings` completion.

**`per_pkg_response == "{}"` string comparison** — Fixed in 2dcc4ce. Replaced with `jq -e '. == {}'` structural check, removing dependency on jq `-c` output format.

**`cat "$OSV_MOCK_FILE" || echo ""` hides cat errors** — Fixed in 2dcc4ce. Now uses `if ! cat "$OSV_MOCK_FILE"; then echo "WARNING: ..." >&2; echo ""; fi`.

**`--quick` cost claim inconsistent (HELP.md still said ~75%)** — Fixed in 2dcc4ce. HELP.md now says "Roughly 60–80% cheaper depending on diff composition" matching README and CLAUDE.md. Also fixed the `--security-only` description in SKILL.md's top-level flags list (was "run security-reviewer only"; now includes CVE check).

**Phase 1c reachability annotation rule backwards** — Fixed in 2dcc4ce. `reachable` findings now get a `[REACHABLE]` prefix (the most actionable, highest-visibility signal); `dev-only` and `transitive-only` get de-emphasis suffixes; `unknown` is unannotated.

**Phase 1c missing `subagent_type`/`model`** — Fixed in 2dcc4ce. Phase 1c now specifies `subagent_type: security-reviewer, model: opus` to match the pattern used for all other agent launches.

**Per-result `jq` O(n) fork** — Won't fix in this PR. The architecture reviewer is correct that this is O(n) jq processes, but for real-world manifests (typically <200 packages) the overhead is negligible vs. the eliminated HTTP round-trips. Tracking as a future optimization.

**`--quick` OSV batch response schema assumed** — Won't fix in this PR. The OSV `/v1/querybatch` spec explicitly guarantees the response shape; adding a runtime schema assertion would add noise for a theoretical API drift scenario. Added a comment cross-referencing the OSV schema in the code.

### Low

**`<D>` placeholder undefined** — Fixed in 2dcc4ce. Now reads: `<D> discarded (ruled out by Pass 2)`.

**edge-case-hunter NONE sentinel** — Not an issue. The `NONE` sentinel instruction is present at line 79 of `agents/edge-case-hunter.md`. The removed "no findings" prose was redundant with it.

File: scripts/run-cve-check.sh [hunk 3]
```diff
@@ -308,19 +381,25 @@ parse_package_json() {
...
+  local file="$1" result
+  # Accept ==, ===, ~=, >=, <= version specs; extract the anchor version.
+  # Unpinned packages (no version spec) are skipped — OSV needs a concrete version.
+  if ! result=$(grep -E '^[[:space:]]*[A-Za-z0-9_.-]' "$file" 2>/dev/null \
+      | grep -v '^[[:space:]]*#' \
+      | sed 's/[[:space:]]*#.*//' \
+      | grep -E '(===?|~=|>=|<=)' \
+      | while IFS= read -r line; do
+          line="${line#"${line%%[! ]*}"}"
+          # Strip extras (e.g. requests[security] -> requests)
```
[the remark this answers is not in the document]
Comment (gchaix) [reply scripts/run-cve-check.sh:397]: Fixed in 9ca7465. Restructured to use a subshell with `set +o pipefail` so `grep` exit-1 (no matching lines — valid for files with no pinned deps) is not misinterpreted as a parse error. The spurious WARNING no longer fires for clean requirements files.

**Comment by gchaix** [comment]
Addressing findings from the follow-on ai-pr-review (review 4149223288), fixed in 9ca7465:

**`parse_requirements_txt` spurious WARNING on grep no-match** (inline, `scripts/run-cve-check.sh:393`) — Fixed. The `if ! result=$(...)` wrapper triggered the WARNING branch whenever `grep` exited 1 (no matching lines), which is the normal case for requirements files with no pinned dependencies. Restructured to use a subshell with `set +o pipefail` so grep no-match is treated as an empty result, not an error.

**`results_count` used in arithmetic without numeric validation** (`scripts/run-cve-check.sh:524`) — Fixed. Added a `[[ "$results_count" =~ ^[0-9]+$ ]]` regex guard before the `-ne` comparison. If `jq` returns `null`, an empty string, or an error message, the guard fails and the comparison is skipped — no bash arithmetic error.
